django/db/models/sql/compiler.py

```diff
@@ -987,8 +987,12 @@ def as_sql(self):
         opts = self.query.get_meta()
         result = ['INSERT INTO %s' % qn(opts.db_table)]
 
-        has_fields = bool(self.query.fields)
-        fields = self.query.fields if has_fields else [opts.pk]
+        # Filter out readonly fields
+        fields = [field for field in self.query.fields if not field.readonly]
```

Comment:
Is there any case where `self.query.fields` can be `None`? The previous code was preventing from iterating over it in this case.